### ISSUE TITLE
Bootstrap 5 installation instructions with webpack + incompatibility with PopperJS v2

### ISSUE THREAD

**Issue by ahukkanen** [DashboardCode/BsMultiSelect#42]
The installation instructions in the readme do not cover clearly how you can use this with webpacker, so I thought to write the instructions for the steps I went through.

1. First, you need to install PopperJS version 1 which is not told in the installation instructions. You can get the multiselect working also with version 2 of PopperJS but you will see a bunch of errors in the console if you try to do that (more on that below).
```bash
npm i popper.js
```

2. Obviously, you also need to install this package
```bash
npm i @dashboardcode/bsmultiselect
```

3. You'll need to initiate the BsMultiSelect factory method in order to attach the input elements with it

```js
import Popper from "popper.js";
import { ModuleFactory as bsMultiSelectFactory } from "@dashboardcode/bsmultiselect";

// in case the library is updated to be compatible with PopperJS version 2, you can replace this with:
// import { createPopper } from "@popperjs/core";
const createPopper = (ref, pop, opts) => new Popper(ref, pop, opts);
const multiSelectEnv = {window, createPopper};
const multiSelect = bsMultiSelectFactory(multiSelectEnv).BsMultiSelect;
```

4. Finally, you can attach the BsMultiSelect to the elements by using the just initiated `multiSelect` factory method

```js
// With vanilla JS
window.addEventListener("DOMContentLoaded", () => {
  document.querySelectorAll("select[multiple]").forEach((select) => {
    multiSelect(select);
  });
});

// Or if you are using jQuery
$(() => {
  $("select[multiple]").each((_i, select) => {
    multiSelect(select);
  });
});
```

If you are trying to use PopperJS version 2, you can import the `createPopper` method directly from the package as described in the readme and the multiselect element will still work but you will see the following errors in the console:

```
validateModifiers.js:62 PopperJS: an invalid property has been provided to the "undefined" modifier, valid properties are "name", "enabled", "phase", "fn", "effect", "requires", "options"; but "preventOverflow" was provided.
validateModifiers.js:62 PopperJS: an invalid property has been provided to the "undefined" modifier, valid properties are "name", "enabled", "phase", "fn", "effect", "requires", "options"; but "hide" was provided.
validateModifiers.js:62 PopperJS: an invalid property has been provided to the "undefined" modifier, valid properties are "name", "enabled", "phase", "fn", "effect", "requires", "options"; but "flip" was provided.
```

This is because BsMultiSelect is not compatible with PopperJS version 2 at this point. This seems to be because of the way the `modifiers` option is passed to the `createPopper` method here:
https://github.com/DashboardCode/BsMultiSelect/blob/dda6c9246aa628ddeeb5e77460590e184d209901/js/src/plugins/CreatePopperPlugin.js#L20-L24

This is not according to the version 2 documentation:
https://popper.js.org/docs/v2/modifiers/

But if you take a look at the version 1 documentation, it seems to be in line with that:
https://popper.js.org/docs/v1/#Popper.Defaults

**Comment by rpokrovskij** [comment]
Thank you for this great report.  I will improve this soon, and give you more details on using BsM swith webpack (to include only those plugins you need).

But one note ahead: 

> const multiSelect = (new BsMultiSelectFactory(multiSelectEnv)).BsMultiSelect;

There should be not need for "new". BsMultiSelectFactory is just a method

**Comment by ahukkanen** [comment]
> There should be not need for "new". BsMultiSelectFactory is just a method

True that. I updated the instructions above.

**Comment by rpokrovskij** [comment]
@ahukkanen 
can you test new version 1.1.16 ?
I was unable to reproduce warning messages, but have changed popper "modifiers" according newest popper documentation.

**Comment by ahukkanen** [comment]
@rpokrovskij

Thanks a lot! I tested 1.1.16 and it works with Popper.JS version 2 without any console errors.

However, the usage instructions are a bit different than what you wrote above. Here is how I got it working:

```js
import { createPopper } from "@popperjs/core";
import { ModuleFactory as bsMultiSelectFactory } from "@dashboardcode/bsmultiselect";

const multiSelectEnv = {window, Popper: createPopper};
const multiSelect = bsMultiSelectFactory(multiSelectEnv).BsMultiSelect;
```

There is no export named `Popper` in the `@popperjs/core` package, so I figured you meant the `createPopper` method.

Also, with Popper.js version 1, there is no `createPopper` export in the package, so you need to create the method yourself as explained in the original post.

As for coding conventions, I think it would be clearer if with version 2 we could provide the `createPopper` method for the environment and with version 1 the `Popper` constructor object would be provided. The current convention is quite confusing and it took me a while to figure it out.

**Comment by rpokrovskij** [comment]
@ahukkanen  Yes you are right I have wrote it wrong I'm sorry

view-source:https://dashboardcode.github.io/BsMultiSelect/snippetEsm.html

import {createPopper} from "https://cdn.jsdelivr.net/npm/@popperjs/core@2.9.3/dist/esm/popper.min.js"


view-source:https://dashboardcode.github.io/BsMultiSelect/snippetEsmBs4.html

 import Popper from "https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/esm/popper.min.js"

I have intentionaly reserved createPopper and Popper property names as `environment` object's  for different Popper versions. This  inconsistency is "inherited" from popper. Different properties is also the "flag" to me : how to create popper's modifiers as for V1 or for V2.

"One moment" I will just stop support popper V1 and the inconsistency will be gone this way.

**Comment by ahukkanen** [comment]
Just FYI, I tested 1.17.0 + I bumped Popper.js to 2.10.0, and now it works as advertised without the initialization inconsistency and without any errors in the console.

When I tested 1.17.0 + Popper.js 2.9.3, I still got the console errors when trying to provide the `createPopper` name to the environment variable.

I'm wondering if the problem was at Popper.js, haven't looked further into it. There were some changes in Popper.js 2.10.0 touching the options definitions but it didn't look like it would be related.

**Comment by rpokrovskij** [comment]
Thank you. Going to experiment with popper.js 2.9.3 .  Could you send me your snipped project with webpack.config

**Comment by ahukkanen** [comment]
Here is my webpack.config.js:

```js
/* eslint-disable no-ternary, no-process-env */

const path = require("path");
const TerserPlugin = require("terser-webpack-plugin");
const PostCSSPresetEnv = require("postcss-preset-env");
const CssMinimizerPlugin = require("css-minimizer-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");
const StylelintPlugin = require("stylelint-webpack-plugin");
const { WebpackManifestPlugin } = require("webpack-manifest-plugin");
const ESLintPlugin = require("eslint-webpack-plugin");

const isDev = process.env.NODE_ENV !== "production";

const esLintOptions = {
  extensions: ["js"],
  exclude: [
    "node_modules",
    "./config",
  ],
}

module.exports = {
  mode: isDev
    ? "development"
    : "production",
  stats: {
    colors: true,
    preset: "minimal",
  },
  ignoreWarnings: [
    {
      file: /bootstrap/,
    },
  ],
  performance: {
    hints: isDev
      ? false
      : "warning",
  },
  devtool: isDev
    ? "cheap-module-source-map"
    : "source-map",
  entry: [
    path.resolve(__dirname, "src/scripts/main.js"),
    path.resolve(__dirname, "src/styles/main.scss"),
  ],
  output: {
    publicPath: "/assets/",
    path: path.resolve(__dirname, "dist/assets"),
    filename: isDev
      ? "[name].js"
      : "[name].[contenthash].js",
  },
  plugins: [
    new ESLintPlugin(esLintOptions),
    new WebpackManifestPlugin(),
    new StylelintPlugin({
      files: "src/styles/**/*.scss",
    }),
    new MiniCssExtractPlugin({
      filename: isDev
        ? "[name].css"
        : "[name].[contenthash].css",
    }),
  ],
  ...(!isDev && {
    optimization: {
      minimizer: [
        (compiler) => {
          Reflect.apply(
            new TerserPlugin({
              terserOptions: {
                compress: {},
              },
            }),
            compiler
          );
        },
        new CssMinimizerPlugin(),
      ],
    },
  }),
  module: {
    rules: [
      {
        enforce: "pre",
        test: /\.scss/,
        loader: "import-glob-loader",
      },
      {
        test: /\.js$/,
        exclude: /node_modules/,
        loader: "babel-loader",
      },
      {
        test: /\.s?css/i,
        use: [
          MiniCssExtractPlugin.loader,
          "css-loader",
          {
            loader: "postcss-loader",
            options: { postcssOptions: { plugins: [PostCSSPresetEnv] } },
          },
          "sass-loader",
        ],
      },
      {
        test: /\.(png|jpe?g|gif|svg)$/i,
        type: "asset/resource",
        generator: {
          filename: `images/${isDev
            ? "[name][ext]"
            : "[contenthash][ext]"}`,
        },
      },
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/i,
        type: "asset/resource",
        generator: {
          filename: `fonts/${isDev
            ? "[name][ext]"
            : "[contenthash][ext]"}`,
        },
      },
    ],
  },
  resolve: {
    alias: {
      // Helpful alias for importing assets
      assets: path.resolve(__dirname, "src/assets"),
    },
  },
};

```

**Comment by rpokrovskij** [comment]
and how you reference BsMultiSelect in your index.js?

**Comment by ahukkanen** [comment]
I have the following dependencies at `package.json` (right now after the updates):

```json
  "dependencies": {
    "@dashboardcode/bsmultiselect": "^1.1.17",
    "@fortawesome/fontawesome-free": "^5.15.3",
    "@popperjs/core": "^2.10.0",
    "bootstrap": "^5.0.1",
    "jquery": "^3.6.0"
  }
```

And of course a lot of dev dependencies, such as webpack, postcss, babel, linters, etc. But I don't think they are relevant regarding this.

**Comment by rpokrovskij** [comment]
Thank you.
Here view-source:https://dashboardcode.github.io/BsMultiSelect/snippetEsm.html popper  Popper.js 2.9.3 but there are no warningns. I hope with your help I will be able to reproduce them...

**Comment by rpokrovskij** [comment]
Got it. There is difference referencing min and not min Popper versions. "Not min" throw those warnings.

**Comment by rpokrovskij** [comment]
solved in 1.1.18
